DTS Support

YunchuWang opened the pull request:
Two new azuremanaged packages for connecting to durable task schedulers. 
This pull request introduces several significant changes to the `azuremanaged` project, including new build configurations, client extensions, and options for the Durable Task Scheduler. The most important changes are summarized below:

### Build Configuration Updates:
* Added a new `allprojects` configuration to set the group and version for all projects in `azuremanaged/build.gradle`.
* Introduced a new build file for the `client` module with necessary plugins, dependencies, and publishing configurations in `azuremanaged/client/build.gradle`.
* Added a new build file for the `shared` module with required plugins and dependencies in `azuremanaged/shared/build.gradle`.
* Updated the settings to include new modules (`client`, `worker`, `shared`) in `azuremanaged/settings.gradle`.

### Client Extensions:
* Created `DurableTaskSchedulerClientExtensions` class to provide extension methods for creating and configuring `DurableTaskClient` instances.

### Client Options:
* Added `DurableTaskSchedulerClientOptions` class to handle configuration options for the Durable Task Scheduler, including methods for creating gRPC channels and handling authentication.

### Token Management:
* Implemented `AccessTokenCache` class to cache access tokens for Azure authentication, ensuring tokens are refreshed before expiration.

### Workflow Adjustment:
* Modified the build validation workflow to remove the `-x sign` flag from the `publishToMavenLocal` task in `.github/workflows/build-validation.yml`.<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

resolves #issue_for_this_pr

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes are added to the `CHANGELOG.md`
* [ ] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

cgillum commented:
In DTS and Python repos, we changed the build process to instead copy the proto definitions locally rather than rely on a submodule. It's nice because working with submodules can be complicated and hard to track with source control. Should we do the same thing here?

YunchuWang commented:
Running sample working:
client curl
![{953C49AC-AAF2-4D66-84E9-499507FB45FD}](https://github.com/user-attachments/assets/5ffaf1fa-cd93-44c2-81ce-24860b4217c4)

sample logs:
2025-01-23 10:38:38.879  INFO 7760 --- [global-thread-0] c.azure.identity.ChainedTokenCredential  : Azure Identity => Attempted credential AzureCliCredential returns a token
2025-01-23 10:38:39.621 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : a99035b4-2a2a-4324-b0e4-28db2866abd2: calling activity 'ValidateOrder' (#0) with serialized input: "'{orderId:123,"
2025-01-23 10:38:39.621 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : The orchestrator has yielded and will await for new events.
2025-01-23 10:38:39.705 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : a99035b4-2a2a-4324-b0e4-28db2866abd2: Activity 'ValidateOrder' (#0) completed with serialized output: false
2025-01-23 10:39:09.701  WARN 7760 --- [       Thread-1] com.microsoft.durabletask                : Received and dropped an unknown {0} work-item from the sidecar.
2025-01-23 10:39:39.709  WARN 7760 --- [       Thread-1] com.microsoft.durabletask                : Received and dropped an unknown {0} work-item from the sidecar.
2025-01-23 10:40:09.707  WARN 7760 --- [       Thread-1] com.microsoft.durabletask                : Received and dropped an unknown {0} work-item from the sidecar.
2025-01-23 10:40:39.708  WARN 7760 --- [       Thread-1] com.microsoft.durabletask                : Received and dropped an unknown {0} work-item from the sidecar.
2025-01-23 10:41:09.720  WARN 7760 --- [       Thread-1] com.microsoft.durabletask                : Received and dropped an unknown {0} work-item from the sidecar.
2025-01-23 10:41:39.717  WARN 7760 --- [       Thread-1] com.microsoft.durabletask                : Received and dropped an unknown {0} work-item from the sidecar.
2025-01-23 10:41:44.814  INFO 7760 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2025-01-23 10:41:44.814  INFO 7760 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2025-01-23 10:41:44.814  INFO 7760 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 0 ms
2025-01-23 10:41:45.275 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : 740edaa7-2483-4933-bd6f-c064e431d9d6: calling activity 'ValidateOrder' (#0) with serialized input: "'{orderId:123,"
2025-01-23 10:41:45.275 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : The orchestrator has yielded and will await for new events.
2025-01-23 10:41:45.362 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : 740edaa7-2483-4933-bd6f-c064e431d9d6: Activity 'ValidateOrder' (#0) completed with serialized output: false
2025-01-23 10:42:15.365  WARN 7760 --- [       Thread-1] com.microsoft.durabletask                : Received and dropped an unknown {0} work-item from the sidecar.
2025-01-23 10:42:16.000 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : 06be66ce-ede9-46ea-b6a0-d305cc304bcd: calling activity 'ValidateOrder' (#0) with serialized input: "{\"orderId\":\"123\", \"amount\":100}"
2025-01-23 10:42:16.000 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : The orchestrator has yielded and will await for new events.
2025-01-23 10:42:16.059 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : 06be66ce-ede9-46ea-b6a0-d305cc304bcd: Activity 'ValidateOrder' (#0) completed with serialized output: true
2025-01-23 10:42:16.059 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : 06be66ce-ede9-46ea-b6a0-d305cc304bcd: calling activity 'ProcessPayment' (#1) with serialized input: "{\"orderId\":\"123\", \"amount\":100}"
2025-01-23 10:42:16.059 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : The orchestrator has yielded and will await for new events.
2025-01-23 10:42:17.128 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : 06be66ce-ede9-46ea-b6a0-d305cc304bcd: Activity 'ProcessPayment' (#1) completed with serialized output: "{\"success\":true, \"transactionId\":\"TXN1737657737098\"}"
2025-01-23 10:42:17.129 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : 06be66ce-ede9-46ea-b6a0-d305cc304bcd: calling activity 'ShipOrder' (#2) with serialized input: "{\"orderId\":\"123\", \"amount\":100}"
2025-01-23 10:42:17.129 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : The orchestrator has yielded and will await for new events.
2025-01-23 10:42:18.187 DEBUG 7760 --- [       Thread-1] com.microsoft.durabletask                : 06be66ce-ede9-46ea-b6a0-d305cc304bcd: Activity 'ShipOrder' (#2) completed with serialized output: "{\"trackingNumber\":\"TRACK1737657738154\"}"


dashboard showing result:

cgillum commented on internal/durabletask-protobuf/lib/java/build.gradle:
```diff
@@ -0,0 +1,52 @@
+}
+
+version = '0.1.0'
+archivesBaseName = 'durabletask-grpc'
```
Rather than creating an archive just for gRPC, could we instead build this directly into an `azureManaged` archive?

YunchuWang replied:
removed this legacy project, this is not used anymore. every projects are depending on proj client to refresh build and package protos.

cgillum commented on internal/durabletask-protobuf/README.md:
````diff
@@ -0,0 +1,52 @@
+If you need to manually update the proto files, you can run:
+
+```bash
+./gradlew downloadProtoFiles
````
Is it possible to provide parameters to this command? For example, the name of a branch to use other than `main`?

cgillum commented on internal/durabletask-protobuf/README.md:
````diff
@@ -0,0 +1,52 @@
+./gradlew downloadProtoFiles
+```
+
+# Durable Task Protobuf Definitions
````
Should this section and the parts below be removed? I assume it was copied from `microsoft/durabletask-protobuf`.

cgillum commented on internal/durabletask-protobuf/LICENSE:
```diff
@@ -0,0 +1,21 @@
+    MIT License
```
Can this file be removed?

cgillum commented on CHANGELOG.md:
```diff
@@ -1,4 +1,5 @@
 ## placeholder
+* Add automatic proto file download and commit hash tracking during build ([#TBD](https://github.com/microsoft/durabletask-java/pull/TBD))
```
```suggestion
* Add automatic proto file download and commit hash tracking during build ([#201](https://github.com/microsoft/durabletask-java/pull/201))
```

cgillum reviewed (commented):
I'm wondering if we can refactor this a bit, so that the only new package we ship is an `azureManaged` package.

cgillum commented on azuremanaged/build.gradle:
```diff
@@ -0,0 +1,79 @@
+    implementation "com.fasterxml.jackson.core:jackson-core:${jacksonVersion}"
+    implementation "com.fasterxml.jackson.core:jackson-databind:${jacksonVersion}"
+    implementation "com.fasterxml.jackson.core:jackson-annotations:${jacksonVersion}"
+    implementation "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:${jacksonVersion}"
```
Do we need to declare these dependencies even though this package doesn't use them?

YunchuWang replied:
removed

cgillum commented on samples/src/main/java/io/durabletask/samples/WebAppToDurableTaskSchedulerSample.java:
```diff
@@ -0,0 +1,307 @@
+    }
+
+    @Configuration
+    static class DurableTaskConfig {
```
It seems like much of this configuration logic can be abstracted away, similar to how we abstracted it away in the .NET implementation that you did. For example, instead of Java developers creating a `DurableTaskGrpcWorker`, they would create a `DurableTaskSchedulerWorker` and things like token handling and adding the task hub name to the gRPC headers would be taken care of internally. Were you planning to do something similar to this for Java?

YunchuWang replied:
updated

cgillum reviewed (commented):
This is great. Thank you for making these changes. I just have a question about the intended scope of this PR.

YunchuWang commented:
pr connecting to production scheduler working 
![{E59FDCE1-99F2-4350-B188-318C6D7076F9}](https://github.com/user-attachments/assets/9df3dd7a-88db-4a3f-a17e-f842de0de5ff)

cgillum commented on settings.gradle:
```diff
@@ -2,7 +2,11 @@ rootProject.name = 'durabletask-java'
+include ":azuremanaged"
+include ":azuremanaged:dtclient"
+include ":azuremanaged:dtworker"
+include ":azuremanaged:shared"
```
Unlike .NET, the Java SDK only has one package that combines client and worker instead of having two packages. Because of this, it should be unecessary for us to have separate client and worker packages for DTS. Instead, we can just have a single `azuremanaged` package. Do you agree?

YunchuWang replied:
agree, updated to one azuremanaged package, "com.microsoft.durabletask.azuremanaged"

cgillum commented on azuremanaged/src/main/java/com/microsoft/durabletask/azuremanaged/DurableTaskSchedulerClientOptions.java:
```diff
@@ -0,0 +1,268 @@
+     * @return A new DurableTaskSchedulerClientOptions object.
+     */
+    static DurableTaskSchedulerClientOptions fromConnectionString(DurableTaskSchedulerConnectionString connectionString, @Nullable TokenCredential credential) {
+        // TODO: Parse different credential types from connection string
```
Any reason not to do connection string credential config as part of this PR? It seems like it's half implemented right now, and I worry that these method overloads will become confusing if we add it later.

cgillum commented on azuremanaged/src/main/java/com/microsoft/durabletask/azuremanaged/DurableTaskSchedulerConnectionString.java:
```diff
@@ -0,0 +1,130 @@
+        
+        // Validate required properties
+        getRequiredValue("Endpoint");
+        getRequiredValue("Authentication");
```
See my note about your TODO for getting the token from the connection string. If we require Authentication to be set but ignore the value, how will users configure their connection string in production?

YunchuWang replied:
updated

cgillum commented on azuremanaged/build.gradle:
```diff
@@ -0,0 +1,116 @@
+
+archivesBaseName = 'durabletask-azuremanaged'
+group 'com.microsoft'
+version = '1.5.0'
```
Is there any way to indicate that this is a preview version of this SDK extension?

cgillum reviewed (commented):
Added a few comments.

Can you also update the CHANGELOG.md file?

cgillum commented on azuremanaged/src/main/java/com/microsoft/durabletask/azuremanaged/DurableTaskSchedulerConnectionString.java:
```diff
@@ -127,4 +137,54 @@ private static Map<String, String> parseConnectionString(String connectionString
+     * @return A TokenCredential instance based on the specified authentication type, or null if authentication type is "none".
+     * @throws IllegalArgumentException If the connection string contains an unsupported authentication type.
+     */
+    public @Nullable TokenCredential getCredential() {
```
Last thing: let's add support for the local-friendly credential types, just like we did for the .NET SDK: https://github.com/microsoft/durabletask-dotnet/pull/396

YunchuWang replied:
updated except visualstudio credential which is not supported https://learn.microsoft.com/en-us/java/api/com.azure.identity.intellijcredentialbuilder?view=azure-java-stable. added intellij credential instead
